src/ast/mod.rs

```diff
@@ -1101,405 +978,162 @@ pub enum Action {
-                *lhs,
-                args.iter().map(|e| e.subst(canon)).collect(),
-                rhs.subst(canon),
-            ),
-            Action::Delete(lhs, args) => {
-                Action::Delete(*lhs, args.iter().map(|e| e.subst(canon)).collect())
+            GenericAction::Delete(ann, lhs, args) => {
+                GenericAction::Delete(ann.clone(), lhs.clone(), args.iter().map(f).collect())
+            }
+            GenericAction::Union(ann, lhs, rhs) => {
+                GenericAction::Union(ann.clone(), f(lhs), f(rhs))
             }
-            Action::Union(lhs, rhs) => Action::Union(lhs.subst(canon), rhs.subst(canon)),
-            Action::Extract(expr, variants) => {
-                Action::Extract(expr.subst(canon), variants.subst(canon))
+            GenericAction::Extract(ann, expr, variants) => {
+                GenericAction::Extract(ann.clone(), f(expr), f(variants))
             }
-            Action::Panic(msg) => Action::Panic(msg.clone()),
-            Action::Expr(e) => Action::Expr(e.subst(canon)),
+            GenericAction::Panic(ann, msg) => GenericAction::Panic(ann.clone(), msg.clone()),
+            GenericAction::Expr(ann, e) => GenericAction::Expr(ann.clone(), f(e)),
         }
     }
-}
-
-impl Display for NormAction {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.to_action())
-    }
-}
 
-impl Display for Action {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{}", self.to_sexp())
+    pub fn subst(&self, subst: &mut impl FnMut(&Leaf) -> GenericExpr<Head, Leaf, Ann>) -> Self {
+        self.map_exprs(&mut |e| e.subst_leaf(subst))
     }
-}
 
-#[derive(Clone, Debug)]
-pub struct Rule {
-    // pub query: Query,
-    // pub actions: Vec<Action>,
-    pub head: Vec<Action>,
-    pub body: Vec<Fact>,
-}
-
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub struct NormRule {
-    pub head: Vec<NormAction>,
-    pub body: Vec<NormFact>,
-}
-
-impl NormRule {
-    pub fn to_rule(&self) -> Rule {
-        Rule {
-            head: self.head.iter().map(|a| a.to_action()).collect(),
-            body: self.body.iter().map(|f| f.to_fact()).collect(),
-        }
-    }
-
-    pub fn globals_used_in_matcher(facts: &Vec<NormFact>) -> HashSet<Symbol> {
-        let mut bound_vars = HashSet::<Symbol>::default();
-        for fact in facts {
-            fact.map_def_use(&mut |var, def| {
-                if def {
-                    bound_vars.insert(var);
-                }
-                var
-            });
-        }
-
-        let mut unbound_vars = HashSet::<Symbol>::default();
-        for fact in facts {
-            fact.map_def_use(&mut |var, def| {
-                if !def && !bound_vars.contains(&var) {
-                    unbound_vars.insert(var);
-                }
-                var
-            });
+    pub fn map_def_use(&self, fvar: &mut impl FnMut(&Leaf, bool) -> Leaf) -> Self {
```

Comment:
It is actually funny that every method in this `impl` are dead code if not `pub`. I left a note here. I think `map_def_use` is referred to in `term.rs` so it might be needed in the near future. All these functions are either from the very original codebase or from the `NormFact` refactor. I'm also happy to just delete all of them.